gh-116322: Add Py_mod_gil module slot

swtaarrs opened the pull request:
This PR adds the ability to enable the GIL if it was disabled at interpreter startup, and modifies the multi-phase module initialization path to enable the GIL when loading a module, unless that module's spec includes a slot indicating it can run safely without the GIL.

[PEP 703](https://peps.python.org/pep-0703/#py-mod-gil-slot) called the constant for the slot `Py_mod_gil_not_used`; I went with `Py_MOD_GIL_NOT_USED` for consistency with gh-104148.

A warning will be issued up to once per interpreter for the first GIL-using module that is loaded. If `-v` is given, a shorter message will be printed to stderr every time a GIL-using module is loaded (including the first one that issues a warning).

- Issue: gh-116322

<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--116882.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

swtaarrs committed:
Add Py_mod_gil module slot

swtaarrs committed:
Merge remote-tracking branch 'upstream/main' into cpython-mod-gil

ericsnowcurrently commented on Doc/c-api/module.rst:
```diff
@@ -411,6 +411,31 @@ The available slot types are:
+      The module is safe to run without an active GIL.
+
+   This slot is only used by Python builds configured with
+   :option:`--disable-gil`, and determines whether or not importing this module
```
```suggestion
   This slot is ignored by Python builds not configured with
   :option:`--disable-gil`.  Otherwise, it determines whether or not importing this module
```

ericsnowcurrently commented on Doc/c-api/module.rst:
```diff
@@ -411,6 +411,31 @@ The available slot types are:
+   .. c:macro:: Py_MOD_GIL_USED
+
+      The module assumes the presence of the global interpreter lock (GIL), and
+      may access global state without synchronization.
```
```suggestion
      The module depends on the presence of the global interpreter lock (GIL), and
      may access global state without synchronization.
```

ericsnowcurrently commented on Doc/c-api/module.rst:
```diff
@@ -411,6 +411,31 @@ The available slot types are:
+
+   .. c:macro:: Py_MOD_GIL_NOT_USED
+
+      The module is safe to run without an active GIL.
```
Note that there is a lot of overlap with `Py_mod_multiple_interpreters` here.

Per PEP 489, it [defaults to `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED`](https://github.com/python/cpython/pull/104148/files#diff-8c2a0fd137780a0ee11c19aae8e46e8490f7667698cd78a5bde466bc995a08c5R315-R317).  That implies a promise that the module relies on module state and doesn't have any static types.

PEP 684 (per-interpreter GIL) added `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED`, which is basically one step further: a promise that the module really doesn't have any thread-unsafe state/objects shared between interpreters.  If the module relies on any third-party linked libraries have thread-unsafe state (e.g. openssl) then the module promises to somehow manage that third-party state in a thread-safe way.

Thus, I see `Py_MOD_GIL_USED` as equivalent to `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED` and `Py_MOD_GIL_NOT_USED` to `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED`.  There is still probably value in having the distinct module slots, but it may help to clarify the relationship in the docs.

swtaarrs replied:
> Thus, I see `Py_MOD_GIL_USED` as equivalent to `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED` and `Py_MOD_GIL_NOT_USED` to `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED`.

They're conceptually similar but orthogonal enough that it's definitely worth having both. A couple of the combinations are odd but ultimately make sense [1]; the one that feels odd enough that we might want to ban it is `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED` with `Py_MOD_GIL_NOT_USED`. The most conservative way to interpret this is that multiple threads within one interpreter can run without a GIL, but only threads from one interpreter can be active at a time (because it doesn't claim to support a per-interpreter GIL). It feels unlikely, but maybe some library has thread-safe global state than can be backed by multiple different namespaces (one per interpreter), where only one namespace can be active at a time.

In practice, it will probably be a simpler explanation: the module should be tagged with `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED`, and whoever tagged it with `Py_MOD_GIL_NOT_USED` mistakenly left that out. I don't think we should assume that, though.

I'm guessing we don't want to support the execution mode of "free-threading within an interpreter, but only one interpreter can be active at a time", so what should we do when loading a module with that configuration? We could keep it simple for now and degrade the subinterpreter slot in free-threaded builds, so we'd refuse to load such a module in a non-main interpreter.

I'll add some wording around the interaction between these two slots, once we decide how to handle the case above.

[1] For example: `Py_MOD_GIL_NOT_USED` and `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED` could mean that the module uses a third-party library that is fully thread-safe but has global state that can't be partitioned by interpreter.

ericsnowcurrently replied:
Hmm, one (or both) of us is misunderstanding this.  Let's see if I can break it down.  The key thing is what users are indicating by using the different slot values.

<details>
<summary>(long analysis)</summary>

----

First of all, in order to use module slots a module must implement PEP 489 multi-phase init.  Let's look at how that relates to various critical features in extension modules.  We focus on features that are impacted by thread-safety and interpreter isolation.  Objects are central to this (along with function execution, of course), but other data is affected too.

|  | single-phase init<BR/>(legacy) | multi-phase init |  | data races | interpreter<BR/>isolation |
| --- | --- | --- | --- | --- | --- |
| module C globals     | allowed  | not allowed   |  | X | X |
| static types               | allowed  | not allowed   |  | X | X |
| other global objects | allowed  | not allowed   |  | X | X |
| external C globals    | allowed   | not allowed? |  | X | X |
| module state            | optional | required        |  | X |  |
| heap types               | optional | required         |  | X |  |
| module functions     | always    | always           |  | X |  |

(Reentrancy is another element to the story, but let's ignore that for now.)

Here's how we deal with data races in the different situations:

|  | single interpreter | legacy<BR/>subinterpreter | isolated subinterpreter<BR>(PEP 684) | single nogil<BR>(PEP 703) | legacy nogil<BR>(PEP 703) | isolated nogil<BR>(PEP 703) |
| --- | --- | --- | --- | --- | --- | --- |
| module C globals     | 1 GIL | 1 GIL | ---       | --- | --- | --- |
| static types               | 1 GIL | 1 GIL | ---       | nogil | nogil | --- |
| other global objects | 1 GIL | 1 GIL | ---       | nogil* | nogil* | --- |
| external C globals    | 1 GIL | 1 GIL | ---       | ---     | ---     | --- |
| module state            | 1 GIL | 1 GIL | N GILs | nogil | nogil | nogil |
| heap types               | 1 GIL | 1 GIL | N GILs | nogil | nogil | nogil |
| module functions     | 1 GIL | 1 GIL | N GILs | nogil | nogil | nogil |

\* nogil will protect objects but cannot protect initializing/changing the global variable that holds an object.

Now let's tie it back to the proposed multi-phase init slots, given the following slot values.

`Py_mod_gil`:

A: `Py_MOD_GIL_USED`  (**default**)
B: `Py_MOD_GIL_NOT_USED`

`Py_mod_multiple_interpreters`:

X: `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED`
Y: `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED`  (**default**)
Z: `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED`

(All the above are used via PEP 489 multi-phase init, which is specified indicate that the module supports multiple interpreters.)

Here's how each module slot value relates to the GIL:

| slot | value | single<BR/>interpreter | legacy | isolated | single nogil | legacy nogil | isolated nogil |
| --- | --- | --- | --- | --- | --- | --- | --- |
| `Py_mod_gil` |  |  |  |  |  |  |
|  | A | GIL | GIL | GIL | GIL | GIL | GIL |
|  | B | nogil | nogil | nogil | nogil | nogil | nogil |
| `Py_mod_multiple_interpreters` |  |  |  |  |  |  |  |
|  | X | shared | shared | --- | nogil | nogil | --- |
|  | Y | shared | shared | --- | nogil | nogil | --- |
|  | Z | shared | shared | own | nogil | nogil | nogil |

Here's how each relates to the module features from above, where "X" means "allowed":

|  | A | B |  | X | Y | Z |
| --- | --- | --- | --- | --- | --- | --- |
| module C globals     |  |  |  | allowed |  |  |
| static types               |  |  |  | allowed |  |  |
| other global objects |  |  |  | allowed |  |  |
| external C globals     | allowed? |  |  | allowed | ??? |  |
| module state            | required | required |  | required | required | required |
| heap types                | required | required |  | required | required | required |
| module functions     | required | required |  | required | required | required |

----

</details>

Based on the above, here are the permutations and what they mean:

| `Py_mod_multiple_interpreters` | `Py_mod_gil` | module state<BR/>isolated | external state<BR/>isolated | module state<BR/>thread-safe | external state<BR/>thread-safe |
| --- | --- | --- | --- | --- | --- |
| \* | \*\* | X |  |  |  |
| \* | `Py_MOD_GIL_USED` | X |  |  |  |
| \* | `Py_MOD_GIL_USED` | X |  |  |  |
|  |  |  |  |  |  |
| `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED` | \*\* |  |  |  |  |
| `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED` | \*\* | X |  |  |  |
| `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED` | \*\* | X | X |  | X |
|  |  |  |  |  |  |
| `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED` | `Py_MOD_GIL_USED` |  |  |  |  |
| `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED` | `Py_MOD_GIL_USED` | X |  |  |  |
| `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED` | `Py_MOD_GIL_USED` | X | X |  | X |
|  |  |  |  |  |  |
| `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED` | `Py_MOD_GIL_NOT_USED` |  |  | X | X |
| `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED` | `Py_MOD_GIL_NOT_USED` | X |  | X | X |
| `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED` | `Py_MOD_GIL_NOT_USED` | X | X | X | X |

\* defaults to `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED`
\*\* defaults to `Py_MOD_GIL_USED`

That's a little bit different from what I had in my head.  `Py_MOD_GIL_NOT_USED` implies at least `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED`.  However, it doesn't actually imply `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED`, because the module may need to manage external (library) state relative to each interpreter.  (That's an uncommon case, but not one we can ignore.)

`Py_MOD_GIL_USED` doesn't imply anything relative to `Py_mod_multiple_interpreters`.

> the one that feels odd enough that we might want to ban it is Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED with Py_MOD_GIL_NOT_USED

There's no harm in it.  It would just keep the module from being imported in an interpreter that has maximum isolation.  (The GIL is only one element of that level of isolation.)

> I'm guessing we don't want to support the execution mode of "free-threading within an interpreter, but only one interpreter can be active at a time", so what should we do when loading a module with that configuration?

I'm not clear on why we would need to do anything.  Could you elaborate?

swtaarrs replied:
Thanks for the thorough response!

I'm having trouble making sense of your last table (the one after the collapsed section). Aren't there only 6 possibilities to consider, 3 (`Py_mod_interpreters`) by 2 (`Py_mod_gil`)? Is your table treating "defaults to `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED`" differently than a module explicitly providing `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED` (and similarly for `Py_MOD_GIL_USED`)?

This is the table I've been using to think about the different options (common prefixes omitted for brevity):
| | `GIL_USED` | `GIL_NOT_USED` |
| --- | --- | --- |
| `MULTIPLE_INTERPRETERS_NOT_SUPPORTED` | uses global state (either in the module itself or a supporting library), not thread-safe | uses global state, but is thread-safe |
| `MULTIPLE_INTERPRETERS_SUPPORTED` | default for multi-phase init (state partitioned by interpreter, not thread-safe) | (discussed below) |
| `PER_INTERPRETER_GIL_SUPPORTED` | state can be accessed by multiple interpreters in parallel, but not more than one thread per interpreter at a time | state can be accessed by multiple interpreters with multiple threads each in parallel |

With that context:

> `Py_MOD_GIL_NOT_USED` implies at least `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED`

I don't think this is true. It's not too hard to imagine a C library that is thread-safe but has a global namespace that can't reasonably be used in a way that doesn't share data between subinterpreters (top right square in my table).

And the "(discussed below)" square:

> > the one that feels odd enough that we might want to ban it is Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED with Py_MOD_GIL_NOT_USED
> 
> There's no harm in it. It would just keep the module from being imported in an interpreter that has maximum isolation. (The GIL is only one element of that level of isolation.)

My concern is that if we load such a module in multiple interpreters, while their shared GIL is disabled, we could end up running that module's code in different interpreters in parallel. That seems like the crucial thing that `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED` disallows compared to `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED`. I chatted with @colesbury about this specific issue earlier today, and he's not concerned about it. If you feel the same way then I'm happy to leave this point alone and move on.

I had been thinking of `Py_MOD_GIL_NOT_USED` as meaning "this module supports free-threading within an interpreter. free-threading across interpreters is determined by the `Py_mod_interpreters` slot." But it sounds like you and Sam have been viewing it as "this module supports free-threading, full stop." Does that framing sound reasonable?

> I'm not clear on why we would need to do anything. Could you elaborate?

This was me working through a more complicated scenario that isn't relevant if the `Py_MOD_GIL_NOT_USED` + `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED` combination is freely allowed.

ericsnowcurrently commented on Objects/moduleobject.c:
```diff
@@ -249,6 +251,8 @@ PyModule_FromDefAndSpec2(PyModuleDef* def, PyObject *spec, int module_api_versio
     int has_multiple_interpreters_slot = 0;
     void *multiple_interpreters = (void *)0;
+    int has_gil_slot = 0;
+    void *gil_slot = (void *)Py_MOD_GIL_USED;
```
Why is this causing warnings?

swtaarrs replied:
Oh, because `gil_slot` is never read in a normal build. I can add a cast to void to fix that.

ericsnowcurrently commented on Objects/moduleobject.c:
```diff
@@ -333,6 +348,27 @@ PyModule_FromDefAndSpec2(PyModuleDef* def, PyObject *spec, int module_api_versio
+#ifdef Py_GIL_DISABLED
+    PyThreadState *tstate = _PyThreadState_GET();
+    const PyConfig *config = _PyInterpreterState_GetConfig(interp);
+    if (gil_slot == Py_MOD_GIL_USED && config->enable_gil == _PyConfig_GIL_DEFAULT) {
```
As noted above, there's a strong relationship between `Py_mod_multiple_interpreters` and `Py_mod_gil`.  I think the code should reflect that here:

```suggestion
    if (has_gil_slot) {
        if (has_multiple_interpreters_slot) {
            if (gil_slot == Py_MOD_GIL_USED && multiple_interpreters == Py_MOD_PER_INTERPRETER_GIL_SUPPORTED) {
                // Warn about this mismatch?
            }
            else if (gil_slot == Py_MOD_GIL_NOT_USED && multiple_interpreters == Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED) {
                // Warn about this mismatch?
            }
        }
        else {
            // Imply?
            multiple_interpreters = gil_slot == Py_MOD_GIL_USED
                ? Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED
                : Py_MOD_PER_INTERPRETER_GIL_SUPPORTED;
    }
#ifdef Py_GIL_DISABLED
    PyThreadState *tstate = _PyThreadState_GET();
    const PyConfig *config = _PyInterpreterState_GetConfig(interp);
    if (gil_slot == Py_MOD_GIL_USED && config->enable_gil == _PyConfig_GIL_DEFAULT) {
```

swtaarrs replied:
I don't think either of those are actually mismatches:
- Line 354: This means that multiple interpreters (with their own GILs) can be running in this module concurrently, but only one thread per interpreter at a time. This could be as straightforward as each interpreter's copy of the module state containing a data structure that isn't thread-safe. In a free-threaded build, we'll already warn the first time any interpreter loads a module like this.
- Line 357: This is a little weirder, but it doesn't seem completely unreasonable to have thread-safe global state that can't be partitioned by interpreter (due to design restrictions of a third-party library, for example).

I'm confused by line 363 (and 351, by extension). My reading of your suggested code is "if the module sets `Py_mod_gil` to `Py_MOD_GIL_USED` and doesn't provide `Py_mod_multiple_interpreters`, default `Py_mod_multiple_interpreters` to `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED`". In your [comment on `module.rst`](https://github.com/python/cpython/pull/116882/files#r1538106137), you said that the default is `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED`. That lines up with what [the code does](https://github.com/python/cpython/blob/0821923aa979a72464c5da8dfa53a719bba5801c/Objects/moduleobject.c#L316-L320), but [the docs say](https://github.com/python/cpython/blob/0821923aa979a72464c5da8dfa53a719bba5801c/Doc/c-api/module.rst?plain=1#L409-L410) that the default is `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED`.

Assuming the default in the code is correct, shouldn't line 363 in your suggested code be `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED`? It would be confusing to change the default for `Py_mod_multiple_interpreters` when `Py_mod_gil` is explicitly set to its default value.

For line 364, we shouldn't make that assumption. I discussed this more in my reply to your `module.rst` comment, but I think `Py_MOD_GIL_NOT_USED` and `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED` is the one combination that we should warn on or refuse to load. It's probably an oversight on the module author's part, and they most likely meant to give `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED`, but the risk of subtle bugs from making that assumption seems too high.

ericsnowcurrently replied:
Yeah, I don't think there's anything to change here.

ericsnowcurrently commented on Objects/moduleobject.c:
```diff
@@ -333,6 +348,27 @@ PyModule_FromDefAndSpec2(PyModuleDef* def, PyObject *spec, int module_api_versio
+    PyThreadState *tstate = _PyThreadState_GET();
+    const PyConfig *config = _PyInterpreterState_GetConfig(interp);
+    if (gil_slot == Py_MOD_GIL_USED && config->enable_gil == _PyConfig_GIL_DEFAULT) {
+        if (_PyEval_EnableGIL(tstate)) {
```
FWIW, this is not where I would look for code that turns the GIL back on.  I'd look in Python/import.c (or maybe Python/importdl.c).  That said, I'm not sure how that code could know whether or not the GIL should be enabled for a module.  It would have basically have to repeat the code you have here or we'd store the result on the module (or module def) and refer to that.

I'm not sure that's worth it and yet calling `_PyEval_EnableGIL()` here still doesn't seem right.

swtaarrs replied:
This does feel like it belongs somewhere more explicitly in the import system, but this function is the most natural place for the functionality. If we stored the result on the module and read it later, we'd risk running the `Py_mod_create` slot before disabling the GIL (down on line 373)–that seems like something we should definitely avoid.

ericsnowcurrently replied:
I'm fine with leaving it like you have it, but would you mind leaving a comment noting that ideally we'd find a way to do this where it more naturally fits (in import.c or importdl.c)?

swtaarrs replied:
Yeah, sounds good.

ericsnowcurrently commented on Objects/moduleobject.c:
```diff
@@ -333,6 +348,27 @@ PyModule_FromDefAndSpec2(PyModuleDef* def, PyObject *spec, int module_api_versio
+                "without the GIL. To override this behavior and keep the GIL "
+                "disabled (at your own risk), run with PYTHON_GIL=0 or -Xgil=0.",
+                name
+            );
```
IIRC, we had some discussion somewhere about how often this warning should be emitted.  Do you recall where that was?  What did we decide there?

swtaarrs replied:
It was [in Discord](https://discord.com/channels/854719841091715092/1176203656400474184/1217556609656553543), where you suggested one warning per interpreter, plus one message for each module when `-v` is given. I assumed that "one warning per interpreter" meant at the time that the GIL was actually enabled, so when `_PyEval_EnableGIL()` returns true.

ericsnowcurrently reviewed (changes requested):
mostly LGTM

I've left a few comments and I have some feedback on the relationship between `Py_mod_multiple_interpreters` and `Py_mod_gil`.

swtaarrs committed:
Some review comments from Eric

swtaarrs committed:
Merge remote-tracking branch 'upstream/main' into cpython-mod-gil

swtaarrs commented on Python/ceval_gil.c:
```diff
@@ -966,6 +966,51 @@ _PyEval_InitState(PyInterpreterState *interp)
+    // 4. Start the world.
+    // 5. Attach the current thread. Other threads may attach and hold the GIL
+    //    before this thread, which is harmless.
+    _PyThreadState_Detach(tstate);
```
As I was working out how to safely *disable* the GIL, I realized that this sequence to enable the GIL isn't safe:
1. Thread A is detached and calls `_PyThreadState_Attach()`.
2. [This call to `_PyEval_AcquireLock()`](https://github.com/python/cpython/blob/42205143f8b3211d1392f1d9f2cf6717bdaa5b47/Python/pystate.c#L1967) is a no-op in thread A because the GIL is disabled.
3. Thread B calls `_PyEval_EnableGIL()` and enables the GIL.
4. Thread A may briefly wait in [`tstate_try_attach()`](https://github.com/python/cpython/blob/42205143f8b3211d1392f1d9f2cf6717bdaa5b47/Python/pystate.c#L1973) depending on the timing of B's stop-the-world, but the damage is done: thread A made it past `_PyEval_AcquireLock()` and the GIL is enabled.
5. Thread A attaches, *without* holding the GIL, and continues on. Disaster ensues.

So this also needs some work.

swtaarrs committed:
Fix enabling the GIL, also support disabling the GIL

More details:
- Fix a race while enabling the GIL by checking if the GIL was enabled between
  a no-op call to `_PyEval_AcquireLock()` and the thread attaching, and trying
  again if it was.
- Enable the GIL before running a module init function, since we can't know if
  it's a single-phase init module that doesn't support free-threading. Look at
  the state of the module after initialization to determine if it's safe to
  disable the GIL again.
- Add `PyModule_SetGIL()`, which can be used by single-phase init modules to
  declare that they support running without the GIL.
- Change `gil->enabled` from a simple on/off switch to a count of active
  requests to enable the GIL. This allows us to support multiple interleaved
  imports that each independently track whether the GIL should remain
  enabled. See the big comment in `pycore_ceval.h` for more details.

swtaarrs committed:
Merge remote-tracking branch 'upstream/main' into cpython-mod-gil

swtaarrs committed:
Fix module size in test_objecttypes

Also fix the module format string for normal builds, since it only contains
pointers (no `size_t`).

swtaarrs committed:
Add missing : in Py_mod_gil documentation

ericsnowcurrently commented on Python/pystate.c:
```diff
@@ -2039,19 +2039,36 @@ _PyThreadState_Attach(PyThreadState *tstate)
-    // XXX assert(tstate_is_alive(tstate));
-    current_fast_set(&_PyRuntime, tstate);
-    tstate_activate(tstate);
+    while (1) {
```
This could also be done with a do-while, which is slightly cleaner (no `break` needed):

```C
    do {
        ...
#ifdef Py_GIL_DISABLED
        ...
#endif
    } while (0);
```

swtaarrs replied:
Can you be more specific about the contents of the loop you had in mind here? `continue` is equivalent to `break` in a `do {} while (0)`, so I don't see any way to restart the body of the loop (apart from a label + `goto`).

(this code has been taken out of this PR but I'm getting it ready for the follow-up PR)

ericsnowcurrently replied:
I meant the same body as the while loop.  You're right that there isn't much difference.  I'm not sure why I was thinking `do...while` was an improvement.  Maybe I missed the `continue` or something.

ericsnowcurrently commented on Python/import.c:
```diff
@@ -1400,23 +1404,63 @@ create_builtin(PyThreadState *tstate, PyObject *name, PyObject *spec)
+        _PyEval_DisableGIL(tstate);
+    }
+}
+#endif
```
Consider moving this to right after `_PyImport_CheckSubinterpIncompatibleExtensionAllowed()`.

ericsnowcurrently commented on Include/moduleobject.h:
```diff
@@ -90,6 +94,13 @@ struct PyModuleDef_Slot {
+#  define Py_MOD_GIL_USED ((void *)0)
+#  define Py_MOD_GIL_NOT_USED ((void *)1)
+PyAPI_FUNC(int) PyModule_SetGIL(PyObject *module, void *gil);
+#endif
```
What is this for?  I don't see it used anywhere.

Also, is this really supposed to go in the limited API?  I wouldn't expect it to be there.

ericsnowcurrently replied:
Presumably this is so single-phase init modules can indicate they do or do not support running without a GIL.  To me, it make more sense to say that all single-phase init modules require the GIL.

swtaarrs replied:
> Presumably this is so single-phase init modules can indicate they do or do not support running without a GIL

Yeah, it's for gh-117526. I put it in this PR because it ended up being such a small change.

> it make more sense to say that all single-phase init modules require the GIL

I don't think we should make that assumption. @colesbury said that when he did a survey of popular numerical computing packages (like NumPy and PyTorch), most of them still used single-phase init. Requiring developers to move to multi-phase init to support free-threading could be an unreasonable burden; there's no reason why global state can't be made thread-safe while remaining global state (and not supporting subinterpreters).

> is this really supposed to go in the limited API?

I was modeling it after the existing subinterpreters slot, but it doesn't have to be in the limited API. I can take it out of the limited API for now and we can discuss adding it back later if a need comes up.

Are you talking about `PyModule_SetGIL()` specifically or everything related to the `Py_mod_gil` slot?

ericsnowcurrently replied:
> > it make more sense to say that all single-phase init modules require the GIL
> 
> I don't think we should make that assumption

Changing a module to multi-phase init takes small effort *if* you isolate the module between interpreters (e.g. don't port static types or switch to module state).  Along with that minimal effort, it should then be sufficient to set the `Py_mod_multiple_interpreters` slot to `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED`.  That's a stopgap that would allow people to use the `Py_mod_gil` slot to indicate they support free-threading (or don't).  Then we wouldn't need to add `PyModule_SetGIL()` to accomplish the same thing.

(That said, encouraging folks to implement multi-phase init without implementing interpreter isolation isn't ideal.  I'm sure @encukou has an opinion. 😄)

> Are you talking about `PyModule_SetGIL()` specifically or everything related to the `Py_mod_gil` slot?

I was talking about `PyModule_SetGIL()` specifically.  Putting the new module slot values in the limited API is fine.

encukou replied:
Opinion? I have a [documented solution](https://docs.python.org/3/howto/isolating-extensions.html#opt-out-limiting-to-one-module-object-per-process) that predates `Py_mod_multiple_interpreters` ;)
Alas, it relies on the GIL...

colesbury replied:
`PyModule_SetGIL()` does not need to be part of the limited API. Unfortunately, the limited API doesn't currently work with free-threading, so there isn't much reason to include it as part of the limited API.

I feel pretty strongly that we should not require package maintainers to rewrite their extensions to use multi-phase init, just to support free-threading. Even without isolating state, it's a bunch of extra work that is not necessary. Many packages have multiple C API modules (NumPy has [~15](https://github.com/search?q=repo%3Anumpy%2Fnumpy%20PyMODINIT_FUNC&type=code)) so it would require a substantial amount of refactoring.

ericsnowcurrently replied:
> I feel pretty strongly that we should not require package maintainers to rewrite their extensions to use multi-phase init,

The level of effort would be minimal:

1. move the content of each init function to a corresponding "module exec" function (drop call to create module object)
2. add a static `PyModuleDef_Slot` var for each, with a `Py_mod_exec` slot
3. add the `Py_mod_multiple_interpreters`, set to `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED`
4. set `def.m_size` to 0 (if negative)
5. update each init function to only `return PyModuleDef_Init(def);` for the corresponding module def

There's no need to do any of the other stuff (heap types, module state, etc.) immediately.

At that point, the only extra step to run on a free-threading build would be to add the `Py_mod_gil` slot, set to `Py_MOD_GIL_USED `.

As to single-phase init modules, we'd always treat them as though they had `Py_MOD_GIL_USED `.

> just to support free-threading.

We want people to switch to multi-phase init regardless.  Single-phase init modules continue to be a pain point for the CPython runtime, as well as limiting options (e.g. reload, subinterpreters) for the extension modules and requiring that they do work that the import machinery could do for them.

Free-threading support would be yet another carrot.

> Even without isolating state, it's a bunch of extra work that is not necessary.

The steps I noted above should be all that's necessary and I would not expect them to take more than a matter of minutes for each module (whether one in a file or many).  Perhaps I've missed something?

> it would require a substantial amount of refactoring.

Likewise, I'm not clear on what substantial refactoring would be necessary.

ericsnowcurrently replied:
My main concern is that we'd be adding `PyModule_SetGIL()` when we already have "one obvious way to do it": multi-phase init with the `Py_mod_gil` slot.  (Furthermore, I'd rather we not make any new accommodations for people to stay on single-phase init.)

If you still have a strong opinion in favor of `PyModule_SetGIL()` then I recommend opening a thread on DPO to get feedback from a wider audience.  I realize time is short for the 3.13 feature freeze; perhaps we can find a middle-ground in the next few days, or convince @Yhg1s to let us add `PyModule_SetGIL()` after beta 1 (assuming we end up adding it).

In the meantime, could we land this PR without `PyModule_SetGIL()`?

ambv replied:
@ericsnowcurrently our concern with *requiring* multi-phase init for free-threading support is effectively crippling the PEP 703 effort. Our latest estimates are that close to 90% of existing extensions on PyPI do not in fact use multi-phase init at this point. We won't be able to get much data from them if we put a big yak shave before they can test free threading.

ericsnowcurrently replied:
If it's just a temporary thing so people can more easily try it out then could we at least call the function something more obvious and less likely to inhibit the move to multi-phase init, like `PyModule_ExperimentalSetGIL()`, and make it available only on `Py_DISABLE_GIL` builds?

colesbury replied:
The `PyModule_ExperimentalSetGIL` name and `Py_DISABLE_GIL` guard sounds fine to me.

swtaarrs replied:
> call the function something more obvious and less likely to inhibit the move to multi-phase init, like `PyModule_ExperimentalSetGIL()`, and make it available only on `Py_DISABLE_GIL` builds?

Yeah, I can definitely do that. There's also the [existing convention for marking C-API functions as unstable](https://docs.python.org/3/c-api/stable.html#unstable-c-api), so another option is `PyUnstable_Module_SetGIL()`. Which one would you prefer?

I'm currently working on the updates described in [this comment](https://github.com/python/cpython/pull/116882#issuecomment-2086475722), and I'll include your suggested change.

ericsnowcurrently replied:
`PyUnstable_` sounds good, but I think it would also still help to make it extra obvious: `PyUnstable_Module_ExperimentalSetGIL()`.  If that's too much then I'd rather `PyModule_ExperimentalSetGIL()`.

ericsnowcurrently reviewed (changes requested):
Mostly LGTM.  I'm not sure we need `PyModule_SetGIL()`.  Otherwise I have some minor comments.

Also, sorry I didn't get gh-118116 up (and merged) sooner.  It would have helped you.

swtaarrs committed:
Merge remote-tracking branch 'upstream/main' into cpython-mod-gil

swtaarrs committed:
From Eric: better loop, move _PyImport_CheckGILForModule

swtaarrs committed:
Merge remote-tracking branch 'upstream/main' into cpython-mod-gil

swtaarrs committed:
Don't put PyModule_SetGIL() in the limited API

swtaarrs committed:
Remove code to enable/disable the GIL (it will go in a different PR)

swtaarrs committed:
Update NEWS entry for behavioral changes

swtaarrs committed:
Mark all modules as not using GIL

swtaarrs committed:
Set m->md_gil in PyModule_FromDefAndSpec2, rename/guard PyModule_SetGIL()

swtaarrs committed:
Fix limited API version for Py_mod_gil

swtaarrs committed:
Merge remote-tracking branch 'upstream/main' into HEAD

ericsnowcurrently commented on Include/internal/pycore_moduleobject.h:
```diff
@@ -22,6 +22,9 @@ typedef struct {
     PyObject *md_weaklist;
     // for logging purposes after md_dict is cleared
     PyObject *md_name;
+#ifdef Py_GIL_DISABLED
+    void *md_gil;
+#endif
```
There are alternatives to adding this to `PyModuleObject`, which would be better long-term.  However, that doesn't need to be addressed in this PR.

ericsnowcurrently commented on Objects/moduleobject.c:
```diff
@@ -261,6 +264,8 @@ PyModule_FromDefAndSpec2(PyModuleDef* def, PyObject *spec, int module_api_versio
     int has_multiple_interpreters_slot = 0;
     void *multiple_interpreters = (void *)0;
+    int has_gil_slot = 0;
+    void *gil_slot = Py_MOD_GIL_USED;
```
Please be sure to address the compiler warnings here.

```suggestion
#ifdef Py_GIL_DISABLED
    void *gil_slot = Py_MOD_GIL_USED;
#endif
```

swtaarrs replied:
I'll add a cast to void instead. I had one in a previous version of the PR but it got lost in some shuffling.

ericsnowcurrently reviewed (approved):
Other than the compiler warning, LGTM.

swtaarrs committed:
Fix 'gil_slot set but not used' warning

swtaarrs committed:
Patch generator for Python-ast.c instead of just the file itself

swtaarrs committed:
Update limited API version for _scproxy.c

swtaarrs committed:
Fix _testconsole.c for Windows build

swtaarrs committed:
Fix winsound.c for Windows build

swtaarrs committed:
Mark more modules as not using the GIL

Yhg1s commented on Doc/c-api/module.rst:
```diff
@@ -609,6 +634,19 @@ state:
 
    .. versionadded:: 3.9
 
+.. c:function:: int PyModule_ExperimentalSetGIL(PyObject *module, void *gil)
```
Ehm, can this please be renamed to PyUnstable as suggested before? We have such naming schemes for a reason.

swtaarrs replied:
Sorry, I'll get a PR ready. I went with this name because having both `Unstable` and `Experimental` in the name felt redundant, and in [this comment](https://github.com/python/cpython/pull/116882#discussion_r1586715045), @ericsnowcurrently expressed a preference for `PyModule_ExperimentalSetGIL()` over `PyUnstable_Module_SetGIL()`.

Does anyone else have strong feelings on whether it should be `PyUnstable_Module_ExperimentalSetGIL()` or `PyUnstable_Module_SetGIL()`?

encukou replied:
PR here: https://github.com/python/cpython/pull/118645

IMO, the “experimental” is redundant. If we need “Experimental” too, we should rethink the entire PEP-689 naming convention.

serhiy-storchaka commented on Objects/moduleobject.c:
```diff
@@ -249,6 +249,9 @@ _PyModule_CreateInitialized(PyModuleDef* module, int module_api_version)
         }
     }
     m->md_def = module;
+#ifdef Py_GIL_DISABLE
```
Should be `Py_GIL_DISABLED`.
